### PR TITLE
Add disk image snapshot support

### PR DESCRIPTION
**Pull request by luohao**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
fixes
**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/924

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Add parameters to VolumeSnapshotClass for disk image config.
```

### CODE REVIEW

File: pkg/common/utils.go [hunk 1]
```diff
@@ -114,15 +114,15 @@ func GenerateUnderspecifiedVolumeID(diskName string, isZonal bool) string {
 	return fmt.Sprintf(volIDRegionalFmt, UnspecifiedValue, UnspecifiedValue, diskName)
 }
 
-func SnapshotIDToProjectKey(id string) (string, string, error) {
+func SnapshotIDToProjectKey(id string) (string, string, string, error) {
 	splitId := strings.Split(id, "/")
 	if len(splitId) != snapshotTotalElements {
-		return "", "", fmt.Errorf("failed to get id components. Expected projects/{project}/global/snapshot/{name}. Got: %s", id)
+		return "", "", "", fmt.Errorf("failed to get id components. Expected projects/{project}/global/snapshot/{name}. Got: %s", id)
```
Comment (mattcary) [inline pkg/common/utils.go:120]: Update error message to something like "expected projects/{project}/global/{snapshots|images}/{name}

**Comment by mattcary** [comment]
/assign @amacaskill 

Alexis, can you look at this? You're most familiar with our volume cloning code.

Note that Han & I agreed offline about removing the paging for listing snapshots (& disk images). We ended up doing a similar thing in the filestore driver. The consumer of ListSnapshots just concatenates all the pages together anyway, so it makes it much easier to combine the disk & snapshot lists if we don't try to maintain the paging too.

File: pkg/common/utils.go (hunk 1 above)
Comment (luohao) [reply pkg/common/utils.go]: Updated, thanks!

File: pkg/gce-pd-csi-driver/controller.go [hunk 2]
```diff
@@ -22,6 +22,8 @@ import (
 	"strings"
 	"time"
 
+	"github.com/golang/protobuf/ptypes/timestamp"
+
```
Comment (amacaskill) [inline pkg/gce-pd-csi-driver/controller.go:26]: remove extra blank line

Comment (amacaskill) [inline pkg/gce-pd-csi-driver/controller.go:26]: remove extra blank line

File: pkg/gce-cloud-provider/compute/gce-compute.go [hunk 3]
```diff
@@ -398,11 +404,23 @@ func (cloud *CloudProvider) insertRegionalDisk(
...
+		if err != nil {
+			return err
+		}
+		switch snapshotType {
+		case common.DiskSnapshotType:
+			diskToCreate.SourceSnapshot = snapshotID
+		case common.DiskImageType:
+			diskToCreate.SourceImage = snapshotID
+		default:
+			return fmt.Errorf("invalid snapshot tyep in snapshot ID: %s", snapshotType)
```
Comment (amacaskill) [inline pkg/gce-cloud-provider/compute/gce-compute.go:417]: nit: invalid snapshot **type**

**Comment by amacaskill** [comment]
Looks good to me! I just had two small nits to fix. /approve

File: pkg/gce-cloud-provider/compute/gce-compute.go (hunk 3 above)
Comment (luohao) [reply pkg/gce-cloud-provider/compute/gce-compute.go]: Fixed.

File: pkg/gce-pd-csi-driver/controller.go (hunk 2 above)
Comment (luohao) [reply pkg/gce-pd-csi-driver/controller.go]: Fixed.

**Pull request merged**